templates/test/ci/prow-aks/patches/aks-agentpool0.yaml

```diff
@@ -5,3 +5,5 @@ metadata:
 spec:
   maxPods: 30
   osDiskType: "Managed"
+  enableUltraSSD: true
+  availabilityZones: ["1", "2"]
```

Comment:
Do all the regions we test support availability zones @jackfrancis @mboersma ? Otherwise this might cause test failures when a region that doesn't support zones is selected

And was this added because zones are required for ultra ssd?

Reply:
Yes to the 2nd question.

@CecileRobertMichon is this where configured test regions lives?

- https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/07fbd76dbd7945edb2ef8af968d4693d6dff35f2/scripts/ci-entrypoint.sh#L46

If so, then northcentralus does not have availability zones (all other regions in the above list do). See:

- https://azure.microsoft.com/en-us/global-infrastructure/geographies/#geographies

Reply:
ok in that case we can't merge this as-is, or tests will fail ~ 1 out of 7 times.

/hold

what can we do here? disable the test or remove the region I guess? Not super into the idea of removing a region since that will 1) increase load on quota from other regions and 2) we will loose coverage for "no AZ" regions (since that does have some different code paths and creates availability sets if zonal redundancy isn't available)

Reply:
IMO having coverage of AZ-dependent functionality is more important than proving that the set of our cluster configuration works on non-AZ-supported regions. The large majority of Azure regions support availability zones, and it is a fundamental building block of any mission critical operational implementation.

Reply:
Interestingly enough, I checked the region in question for support of Ultra Disk Storage, and this page claims North Central US is compatible. https://azure.microsoft.com/en-us/global-infrastructure/services/?regions=us-north-central%2cnon-regional&products=all

As is every location you all have listed in `REGIONS`; I checked them all for Ultra Disk Storage support, and the field is checked in each location's Products Page.

Reply:
This doc https://docs.microsoft.com/en-us/azure/virtual-machines/disks-enable-ultra-ssd?tabs=azure-portal#ga-scope-and-limitations states:

> If a region in the following list lacks availability zones that support ultra disks, then a VM in that region must be deployed without infrastructure redundancy in order to attach an ultra disk.

So it's not that a region MUST support availability zones in order to use ultra disk storage, but it cannot use any other redundancy options (eg. availability sets). So in north central US, ultra SSD is available as long as VMs are single VMs. 

Could you please try to create an AKS cluster in north central US with ultra SSD enabled and see what happens? I wonder if we can get away with simply not listing zones explicitly (ie. does AKS do any sort of defaulting magic here that would place the VMs in zones whenever they are available?). Also wondering if VMSS counts as a "redundancy option", since I believe AKS agent pools are by default VMSS.

Reply:
> does AKS do any sort of defaulting magic here that would place the VMs in zones whenever they are available

According to docs, that isn't the case: https://docs.microsoft.com/en-us/azure/aks/availability-zones#limitations-and-region-availability

It also confirms north central US is not one of the supported regions for availability zones in AKS.

Reply:
Before adding AZ config to the AKS PUT request, we got this error:

E0317 19:48:53.440285       1 controller.go:317] controller/azuremanagedcontrolplane "msg"="Reconciler error" "error"="error creating AzureManagedControlPlane capz-e2e-oxq9h5/capz-e2e-oxq9h5-aks: failed to reconcile AzureManagedControlPlane service managedclusters: failed to create managed cluster, failed to begin operation: Code=\"ErrorCodeEnableUltraSSD\" Message=\"Availability zone is required for UltraSSD setting.\"" "name"="capz-e2e-oxq9h5-aks" "namespace"="capz-e2e-oxq9h5" "reconciler group"="infrastructure.cluster.x-k8s.io" "reconciler kind"="AzureManagedControlPlane" 

So I concluded that we can't enable UltraSSD when we create an AKS cluster without declaring availability zones configuration. (And thus we can't test this new capz AKS functionality without AZ config, either)

Reply:
@TheNakedZealot let's remove northcentral US from the list of test regions for now and add a comment above the list that all test regions must support AvailabilityZones